Phantom go

Syor opened the pull request:
Add the game of Phantom Go based on Go Tagging @michalsustr for review.
Creating new request beacuse of failing CLA in #743 pull request.

lanctot commented:
Hi @Syor, 

Do you see this panel?
![image](https://user-images.githubusercontent.com/3929263/143298646-f9ac47f3-c98f-4528-b89b-9ba0fd8c72fc.png)

If so, can you click on "Details" to the right of cla/google "X Missing CLA from one or more contributors" and follow the instructions at that link?

Syor commented:
Hello @lanctot,
Thank you for helping me with the CLA. I found out that I made this pull request also with my school e-mail. Following the instructions, I signed it also on my school e-mail. Should I write some comment tagging the bot to refresh the CLA check? I am not sure if something has changed. 
![image](https://user-images.githubusercontent.com/46754705/143302518-48062b28-209d-4374-906a-8415a7729bb8.png)

lanctot commented:
Great, yes now try tagging the bot like last time?

Syor commented:
@googlebot I signed it

lanctot commented:
Seems to have not worked. Is it possible to open a PR using your github account directly?

lanctot commented:
A ha! I figured it out. Things have changed in how the CLA works and it looks like now I have to manually re-run it a different way (the googlebot seems to no longer exist). I will start a run of the tests now.

lanctot commented:
Ok `pyspiel_test` failed. 

When adding a new game, you need to add it to the list here: https://github.com/deepmind/open_spiel/blob/c3d50a3fb3de7846bcba632619649b9f5d2f284c/open_spiel/python/tests/pyspiel_test.py#L30

Syor commented:
Hello @lanctot,
Thank you for your help with the CLA. 
About the pyspiel_test, I have added the game in my local project, but I forgot to add it to the commited files, I will fix it in next commit.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+    int player_id, std::function<double()> rng) const {
+    int boardSize = board_.board_size();
+
+    std::shared_ptr<const Game> newGame = LoadGame("phantom_go");
```
Use `GetGame`

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+    int boardSize = board_.board_size();
+
+    std::shared_ptr<const Game> newGame = LoadGame("phantom_go");
+    std::unique_ptr<PhantomGoState> newState = std::make_unique<PhantomGoState>(PhantomGoState(newGame, boardSize, komi_, handicap_));
```
and then NewInitialState

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+    {
+        if (infoState[i] != GoColor::kEmpty)
+        {
+            newState->board_.PlayMove(VirtualPointFromBoardPoint(i, boardSize), infoState[i]);
```
Use State::ApplyAction, so that the history is properly updated.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+    return newState;
+}
+
+std::string PhantomGoState::InformationStateString(int player) const {
```
1) This is not a correct implementation -- it does not use imperfect information of the game correctly.
2) Modelling infostates might be too difficult and maybe we should stick to observations only for the moment.

Syor replied:
This method was implemented in the Go game, I havent changed it to the state of Phantom Go: I will comment it for now.

michalsustr replied:
The implementation is not correct -- it's better to remove it than to have misleading implementation. 
Also set `provides_information_state_string=false`

Syor replied:
As "usual", this is a method that remained here from the implementation of Go, on which i started. I will remove the method.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+  return HistoryString();
+}
+
+std::string PhantomGoState::ObservationString(int player) const {
```
Similar as for InformationStateString, this is not a correct implementation.

michalsustr replied:
I don't think this is correct yet either: only board is displayed as observation, but there are other messages from the referee (such as invalid move).

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+}
+
+void PhantomGoState::ObservationTensor(int player, absl::Span<float> values) const {
+  SPIEL_CHECK_GE(player, 0);
```
This method does not use the imperfect info

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,301 @@
+}
+
+void PhantomGoState::UndoAction(Player player, Action action) {
+  // We don't have direct undo functionality, but copying the board and
```
Observations should be popped as well, or maybe we should not implement this method at all.

michalsustr commented on open_spiel/games/phantom_go.h:
```diff
@@ -0,0 +1,173 @@
+  bool superko_;
+};
+
+// Game object.
```
unnecessary comment, remove

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -163,53 +194,28 @@ char GoColorToChar(GoColor c) {
-      ss << "\n";
-  }
-  ss << "   ABCDEFGHJ\n";
+  ss << board_.observationToString();
```
Nice refactor! :)

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.cc:
```diff
@@ -313,31 +314,17 @@ void GoBoard::Clear() {
 }
 
-bool GoBoard::PlayMove(VirtualPoint p, GoColor c) {
+/*void PhantomGoBoard::addEnemyStoneIntoObservation(int boardPoint, int player_id) const {
```
Remove commented out code.

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.cc:
```diff
@@ -481,9 +419,45 @@ void GoBoard::SetStone(VirtualPoint p, GoColor c) {
   board_[p].color = c;
 }
 
+std::array<GoColor, kMaxBoardSize* kMaxBoardSize> PhantomGoBoard::getObservationByID(int player_id) const
```
Use capital letters for method names: https://google.github.io/styleguide/cppguide.html#Function_Names

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.cc:
```diff
@@ -481,9 +419,45 @@ void GoBoard::SetStone(VirtualPoint p, GoColor c) {
+    std::stringstream ss;
+    ss << "\nObservation white:\n";
+
+    for (int x = board_size_ - 1; x >= 0; x--)
```
Please use google code style, you can set up in clion as https://blog.jetbrains.com/clion/2015/07/new-clion-1-1-eap-select-and-use-your-favourite-code-style/

michalsustr replied:
This should reformat
```
for(...) 
{
  ...
}
```
to
```
for(...) {
  ...
}
```

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.h:
```diff
@@ -239,6 +243,15 @@ class GoBoard {
+
+  // On index 0 is stored count of black stones, on index 1 is stored count of white stones
+  // so it equals the enum of GoColor, where kBlack is 0
+  std::array<int, 2> stoneCount;
```
stone_count_

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.h:
```diff
@@ -239,6 +243,15 @@ class GoBoard {
 
+
+  // In this context, GoColor::kEmpty suggests, that a player does not know, what piece is on that exact spot
+  std::array<std::array<GoColor, kMaxBoardSize* kMaxBoardSize>, 2> observations;
```
observations_

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.h:
```diff
@@ -111,18 +113,20 @@ class Neighbours4 {
+  
 
-  std::pair<int, int> stoneCount;
+  std::array<int, 2> getStoneCount() const { return stoneCount; };
```
Rewrite as using the color enum, and then you don't need to use the (uint8_t) casting everywhere.

michalsustr replied:
And add checks for the color being either black or white

michalsustr commented:
Hi @Syor, I made a pass through the code.

lanctot commented:
@Syor can you let me know when you are finished implementing this and when it's ready to import?

Syor commented:
Hello @lanctot, I will let you know as soon as I finish fully implementing this game. I have to implement ResampleFromInfostate to be able to use is-mcts on this game, and to reimplement few methods to match the imperfect information.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -29,84 +29,178 @@ namespace {
+                break;
+            } else {
+                //std::cout << "random action" << ActionToString(c, action) << " was unacceptable\n";
+                newState->UndoAction(-1, -1);
```
Avoid using "magic" numbers like -1 whenever possible, and use the right way to call the function -- the last player and last action, in case someone update the UndoAction() in the future to not use the rollouts.

Syor replied:
I am sorry for this, as i was aware that the implementation in this game does not take in donsideration these values, i just typed in some numbers, not considering possible update of the implementation.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -29,84 +29,178 @@ namespace {
 
-// this method is in progress of making, the implementation is not correct
-std::unique_ptr<State> PhantomGoState::ResampleFromInfostate(
+std::unique_ptr<State> PhantomGoState::ResampleFromInfostateFull( //still to fix moves into eyes that keep messing up the histories
```
This code seems complicated to review. Please use the feedback from the other method to improve your code and simplify it, and I can review it in the next iteration.

Syor replied:
To be honest, it is unfortunately pretty long, considering all the processes to be done to resample from Infostate. I will try to simplify as soon as the method is fully functinal. The hopefully last problem is with the random moves done by the opponent, that might create eyes on the points that the player will play in the future.

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,231 @@
+
+void BasicGoTests() {
+  GameParameters params;
+  params["board_size"] = GameParameter(9);
```
I think you do not need to supply these if they are the default values.

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.h:
```diff
@@ -0,0 +1,324 @@
+
+  
+
+  std::array<int, 2> getStoneCount() const { return stone_count_; };
```
Change the method to `GetStoneCount(GoColor color)` so you avoid the uint8 casting.

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,231 @@
+  }
+}
+
+void ResampleFromInfostateVisualTest()
```
This "test" does not test anything -- there are no checks. Perhaps add a check the observation strings are identical?

Syor replied:
I am aware of the fact, that this is not the test that should be done. This is an implementation that I used during the development of the Metaposition Resampling and I still use it when implementing the Infostate Resampling. The reason that I used this visual test and not a classical test is that visually comparing resampled state and the observation is much more precise to find bugs than using a method to compare them. As soon as both the methods are finished, i will remove this method from the tests.

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,231 @@
+        LoadGame("phantom_go", params);
+    PhantomGoState state(game, kBoardSize, kKomi, 0);
+
+    for(int i = 0; i < 150; i++)
```
Why this constant?

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,231 @@
+    for(int i = 0; i < 150; i++)
+    {
+        std::vector<Action> actions = state.LegalActions();
+        std::shuffle(actions.begin(), actions.end(), std::mt19937(std::random_device()()));
```
You can use SampleAction or even better https://www.cplusplus.com/reference/random/uniform_int_distribution/

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -29,84 +29,178 @@ namespace {
+
+}
+
+std::unique_ptr<State> PhantomGoState::ResampleFromInfostate( //still to fix moves into eyes that keep messing up the histories
```
See #755

Syor commented on open_spiel/games/phantom_go.cc:
```diff
@@ -29,84 +29,178 @@ namespace {
+    for(long action : stones[opp_player_id])
+    {
+        newState->ApplyAction(action);
+        newState->ApplyAction(action);
```
I think it is. The first Aplying is by the opponent, the second is by the player to add the action/stone to his observation.

Syor commented on open_spiel/games/phantom_go.cc:
```diff
@@ -29,84 +29,178 @@ namespace {
+        newState->board_.getStoneCount()[1] == stoneCount[1]))
+    {
+        ToString();
+        newState->ToString();
```
Yes i wanted to. Should I rewrite it using the EQ functions from the Framework?

Syor commented:
I am sorry for this stupid mistake, i thought the deletion of a file would delete it from the commit. I reverted the files into their state before my accidental push of my local implementations.

lanctot commented:
You will probably also need to pull in changes from master (from an hour ago).

lanctot commented:
Hi @Syor @michalsustr, we're discussing a (v1.1) release mid-Feb to early March. Wondering if this game will be ready to be imported before then? Would be cool to include it in the release, if possible.

Syor commented:
Hello @lanctot. If everything goes right, I should finish my winter semester exams by 2. February. There are two "things" that are to be implemented, first and easier is to add Observer, the harder one is to implement Information State sampling. Right now, I would say that the resampling method is about 70% done, right now I am stuck on some enemy stone placements. So if the remaining part of the resampling isn't too difficult, I think I can finish it until the release.

lanctot commented:
Hi @Syor, just an update: we are planning a release for next week, Feb 28th. 

If you do not finish the game before that, I will do another minor release soon after it is finished and merged.

lanctot commented:
Alternatively, we can import the game without the ResampleFromInfoState function (which you can add later) if the rest is ready; that way it can make it into the 1.1 release.

Syor commented:
Hello @lanctot. With @michalsustr we have decided to use OT-Tools CSP solver to make the resampling function. Because of that, the Phantom Go game should be ready to be released in the upcoming release. In the few next days, I will get a code review according to which I will refactor the code to releasable form. As soon as the final touches are commited, I will let you know.

lanctot commented:
Ok cool. This means you need OR-Tools installed to use it?

I would love to have the resampling function conditional on the ORTOOLS build macro: if it is included, then it uses the fancy CSP solver. Otherwise, it is just empty and throws an error. 

That way we can still include it in the list of games by default.

Syor commented:
Hello @lanctot, I am sorry that I forgot to add one important information, which is that I will make another pull request for the resampling function, so the game will be released without it. The problem is pretty big and I am still working on it.

lanctot commented:
Hi @Syor, the CI tests failed. Can you look into it? I cannot import it until they pass. Thanks.

Syor commented:
Hello @lanctot , I am, unfortunately, completely clueless on how to fix these errors, but I will try to fix them somehow. I am sorry if I am not able to make the checks pass until the release. If I will be unsuccessful, I will hopefully fix them with some help from @michalsustr , who helps me with this project.

lanctot commented:
No rush. will probably release tomorrow or in the next few dsys regardless, and we can do another minor release when this is merged.

lanctot commented:
From what I can tell from the output it looks like there are change markers (from git merges) still in is_mcts_test.cc. Are you able to find the CI tests output?

Syor commented:
These are the two error messages.
![image](https://user-images.githubusercontent.com/46754705/155895173-3c1c7529-730e-417b-9867-67a23daca84d.png)
![image](https://user-images.githubusercontent.com/46754705/155895185-c96d18b2-2c2f-417a-b184-024390ccdc95.png)
I have commited few changes by accident, that also included the is_mcts_test.cc file, but I thought I have fixed it. Unfortunately I am not aware of anything that could cause a problem, since I reverted the file into the original form.

Syor commented:
As I am looking at it, there are many changes in the file. I will make a commit with the original file, hopefully it will fix it.

Syor commented:
Just to clarify the previous commit comment, the "residual" code is the Infostate resampling method, that is not finished, but I would like to keep it in the file, in case of possible reuse or completion-

lanctot commented:
Ok now you need to generate a playthrough: https://github.com/deepmind/open_spiel/blob/master/open_spiel/scripts/generate_new_playthrough.sh

michalsustr commented on open_spiel/algorithms/is_mcts_test.cc:
```diff
@@ -28,81 +28,81 @@ namespace {
-              .first;
-    } else {
-      chosen_action = bot->Step(*state);
+    std::unique_ptr<State> state = game.NewInitialState();
```
Please revert changes that only do whitespace formatting.

(i.e. this entire file)

michalsustr commented on open_spiel/games/CMakeLists.txt:
```diff
@@ -345,6 +349,11 @@ add_executable(go_test go_test.cc ${OPEN_SPIEL_OBJECTS}
                $<TARGET_OBJECTS:tests>)
 add_test(go_test go_test)
 
+#new
```
Remove comment "new"

michalsustr commented on open_spiel/games/phantom_go.h:
```diff
@@ -0,0 +1,179 @@
+  }
+  std::vector<Action> LegalActions() const override;
+
+  std::array<int, 2> getStoneCount() const;
```
Rename `GetStoneCount` (consistent naming)

michalsustr commented on open_spiel/games/phantom_go.h:
```diff
@@ -0,0 +1,179 @@
+
+  bool IsTerminal() const override;
+
+  std::unique_ptr<State> ResampleFromMetaposition(
```
Please add comment about what is a metapostion.

michalsustr commented on open_spiel/games/phantom_go.h:
```diff
@@ -0,0 +1,179 @@
+  std::unique_ptr<State> ResampleFromInfostate(
+      int player_id, std::function<double()> rng) const;
+
+  std::unique_ptr<State> ResampleFromMetapositionHard(
```
Why is there a second method for sampling Metapositions?

Syor replied:
This method is used, if the ResampleFromMetaposition fails. This can happen if the count of the stones and the situation on the board and in the players observation is in a such coincidence, that randomly picking remaining enemy stones can result in a such situation, that no more enemy stones can be placed. We have agreed, that if the Metaposition resampling is a temporary solution, because Infostate resampling will be implemented, such solution to this problem is applicable.

michalsustr commented on open_spiel/games/phantom_go/phantom_go_board.h:
```diff
@@ -0,0 +1,305 @@
+Action VirtualActionToAction(int virtual_action, int board_size);
+int ActionToVirtualAction(Action action, int board_size);
+
+/*std::string ActionToString(Action action, int board_size) {
```
Remove commented out code

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,198 @@
+
+    std::cout << "Resampled state\n " << resampleState->ToString();
+
+    /*for(int i = 0; i < state.FullHistory().size(); i++)
```
Remove unused code

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,198 @@
+
+    std::unique_ptr<State> resampleState = state.ResampleFromMetaposition(0, nullptr);
+
+    std::cout << "Original state\n" << state.ToString();
```
This "test" does not properly test anything -- please make sure that the resampled state is indeed a metaposition.

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,198 @@
+    }
+}
+
+//This is a test, that was used to visually analyze resampling
```
// This is a test, that was used to visually analyze resampling.

-- use proper English sentences, with one space after comment //

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,198 @@
+    open_spiel::phantom_go::ConcreteActionsAreUsedInTheAPI();
+    open_spiel::phantom_go::IllegalMoveTest();
+    open_spiel::phantom_go::StoneCountTest();
+    //open_spiel::phantom_go::ResampleFromInfostateVisualTest();
```
Enable tests

Syor replied:
First, I have removed the Visual Test, since it is not an actual test. Second, I think the Force test should be disabled by default, because it is designed to test the resampling on a large number of simulations, and can take several minutes to complete.

michalsustr commented on open_spiel/games/phantom_go_test.cc:
```diff
@@ -0,0 +1,198 @@
+
+//This test was used to test metaposition resampling on large ammounts of states
+//  with different lengths
+void ResampleFromInfostateForceTest() {
```
This is not a proper test, and does not resample from infostate!

lanctot commented:
We released this morning, btw, so no rush on this. We have another game internally in a similar state, so when this is merged we can do a 1.1.1 release for those two games.

Syor commented:
Hello @lanctot, after the current push, I hope that all of the tests should pass. If all of the tests will be ok, I will probably make a final push after a code review in less than a week.

lanctot commented:
Great! Looks like it worked. Will take a look soon.

Syor commented:
Just to clarify, by the code review, I meant a review by @michalsustr, who is a supervisor for my project based on Phantom Go, and makes a code review after every major push to help me maintain the code in publishable form.

lanctot commented:
No problem, we'll take a look after him. Either way, we'll have to look at it before it's merged, but it definitely saves us time to have others helping with that. Thanks @michalsustr !

michalsustr commented on open_spiel/algorithms/is_mcts_test.cc:
```diff
@@ -27,7 +27,7 @@ namespace {
 constexpr const int kSeed = 93879211;
 
-void PlayGame(const Game& game, algorithms::ISMCTSBot* bot, std::mt19937* rng) {
+void PlayGame(const Game &game, algorithms::ISMCTSBot *bot, std::mt19937 *rng) {
```
@Syor Please go over the changed files and revert all unncessary whitespace changes.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,793 @@
+    return state;
+}
+
+//This method is unfinished, will be later replaced by or-tools CSP solver implementation
```
Delete unused code.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,793 @@
+
+//This method is used, when the Metapositon Resampling fails
+//It resamples the state into a Metaposition, that corresponds to the actual state on the game board
+std::unique_ptr<State> PhantomGoState::ResampleFromMetapositionHard(
```
Do you want to keep these implementations? I thought you are doing it via the CSP and this is obsolete.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,793 @@
+
+      if (iig_obs_type_.private_info == PrivateInfoType::kSinglePlayer) {
+          {
+              auto out = allocator->Get("player_observation", {totalBoardPoints});
```
Please encode the board as 3d array: first two dimensions for the board, and 3rd dimension to distinguish between the players and empty cells.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,793 @@
+      if (iig_obs_type_.public_info) {
+
+          {
+              auto out = allocator->Get("history-turns", {state.History().size()});
```
Using the history will result in observations not having a fixed tensor size, which would be better to avoid (it brings little benefit in this case and working with tensors that have dynamics size is more difficult).

michalsustr replied:
You can use only a two-valued observation like `player_turn` to specify which player is currently taking the turn.

michalsustr commented on open_spiel/games/phantom_go.cc:
```diff
@@ -0,0 +1,793 @@
+              std::shared_ptr<const Game> game = state.GetGame();
+              std::unique_ptr<PhantomGoState>
+                  currState = std::make_unique<PhantomGoState>(down_cast<PhantomGoState>(*game->NewInitialState()));
+              auto out = allocator->Get("history-turns", {state.History().size()});
```
Choose a different field name. `history-turns` is already used.

michalsustr replied:
I believe this should encode the stone count differences between turns? Can you encode only the current difference?

michalsustr reviewed (changes requested):
I made a partial review, I can take a closer look once the code is updated. Also some of my previous comments have been marked as resolved, but I don't think they are.

michalsustr commented on open_spiel/algorithms/is_mcts_test.cc:
```diff
@@ -57,7 +57,7 @@ void ISMCTSTest_PlayGame(const std::string& game_name) {
   auto evaluator =
       std::make_shared<algorithms::RandomRolloutEvaluator>(1, kSeed);
 
-  for (algorithms::ISMCTSFinalPolicyType type :
```
Please revert whitespace formatting on the file.

Syor commented:
Closing this pull request. I will create a new one, which will have a better push history.
